Add withCredentials to BrowserDriver

nondeterministic opened the pull request:
I was having trouble getting the browser to send received cookies back to a server, and figured out it was due to RosHTTP not supporting `XMLHttpRequests`'s `withCredentials`-method. So I rudimentarily patched RosHTTP to see if it works, and it does. 

I created this PR, not with the expectation that it could be merged immediately, but to find out whether or not this feature could be added this way, and whether you think it useful.

See also my problems with this missing features as explained in https://stackoverflow.com/questions/52907346/sessions-in-scala-js-and-play

Thanks!

nondeterministic committed:
Add withCredentials to BrowserDriver

hmil commented:
Hi, that's a very good idea. I wasn't aware of this API but I reckon this feature should be present in RösHTTP. Thanks for the investigation.

A couple more things are required to add this feature:
- A test case
- Some documentation in the readme

If you are willing to add these, I'm willing to merge your changes and publish an updated version.

nondeterministic commented:
Great! Will update the PR, when I have pushed this a bit further then.

nondeterministic committed:
Working cookie test for Selenium

nondeterministic committed:
Adding some docu

nondeterministic commented:
Hi again. I have pushed the PR a bit further in the last couple of days. Here's what I've committed:

- Code and README docu added (also replaced the misleading Cookie-header-example with User-Agent in existing docu).
- Test code added to both test server and the Scalas-side of things.
- Reordered contents of `project/plugins.sbt` a bit, because according to documentation and experiments, the browser driver is otherwise not found by uTest.

There are still two problems, however, to which I'd like your input, please:

- The tests are failing, probably because I added additional npm packages (cookie-parser, cors) that are seemingly not installed automatically.  Since I know next to nothing about node.js, could you let me know if I should add a kind of dependencies file, etc.?  (My git diff shows changes in `test/server/package.json` and `test/server/npm-shrinkwrap.json`, but these were so substantial that I hesitated to simply commit them.)

- The code tests fine on my machine using Selenium with the Gecko driver. However, the JVM and JS tests fail for the cookies since obtained cookies are never sent back to the server, and I think this cannot be fixed (see also this answer on SO, which I've found after some intense searching for the causes: https://stackoverflow.com/a/2932428). How should one proceed in this case? Can some tests be skipped in JS and JVM mode but not for Selenium? 

Many thanks!

hmil commented on README.md:
```diff
@@ -137,6 +137,14 @@ HttpRequest("long.source.of/data")
+### Cookies
+
+Servers may use cookies to authenticate users or exchange other information. To make sure that the browser
+adds a previously obtained cookie to subsequent requests of the client, `withCredentials` is offered:
```
This is not entirely true. `withCredentials` only affects cross-domain requests. As stated on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials), "Setting withCredentials has no effect on same-site requests."

hmil commented on shared/src/test/scala/fr/hmil/roshttp/HttpRequestSpec.scala:
```diff
@@ -654,5 +654,25 @@ object HttpRequestSpec extends TestSuite {
+            assert(res.statusCode == 200)
+          })
+      }
+    }
```
These tests aren't isolated, and that will cause flakiness.

Instead of two, you should create a single test case which:
1. makes a first request to set a cookie
2. makes a second request to verify the cookie.

For added security, you should generate a unique ID in the test case, and use this ID in the cookie name so that you isolate the test from potential side effects of cookies surviving multiple test runs.

nondeterministic replied:
DONE.

hmil commented on test/server/index.js:
```diff
@@ -160,6 +162,31 @@ app.post('/streams/in', function(req, res) {
   })
 });
 
+app.get('/send_cookie', function(req, res) {
```
Change the name of this endpoint to "set_cookie", accept a query parameter which shall be the name of the cookie to set.

nondeterministic replied:
DONE.

hmil commented on test/server/index.js:
```diff
@@ -160,6 +162,31 @@ app.post('/streams/in', function(req, res) {
+  }
+});
+
+app.get('/receive_cookie', function(req, res) {
```
Change the name of this endpoint to "verify_cookie". As for the other, add a query parameter which shall be the name of the cookie you expect to find.

nondeterministic replied:
DONE.

hmil commented on shared/src/main/scala/fr/hmil/roshttp/HttpRequest.scala:
```diff
@@ -24,6 +24,7 @@ final class HttpRequest  private (
     val port: Option[Int],
     val protocol: Protocol,
     val queryString: Option[String],
+    val credentials: Boolean,
```
This name doesn't describe the effect very well. Change that to something like `crossDomainCookies`.

nondeterministic replied:
DONE.

hmil reviewed (changes requested):
Yes you need to commit package.json and package-lock.json. It's normal that the changeset in package-lock.json gets huge.

Although this feature maps directly to the underlying XMLHttpRequest's `withCredentials` feature, I think it should have a different name than "withCredentials". That name poorly describes what actually happens and it is a shame that this is what came out as the standard.

Fortunately, RosHTTP is an abstraction layer, and we do not have to follow the naming conventions of the underlying driver being used (as a matter of fact, the underlying driver could be swapped out anytime without a fundamental change of RosHTTP's API).

Therefore, I think this feature should be called "crossDomainCookies", or something similar, and the doc should state that it uses XMLHttpRequest.withCredentials to disable the security restriction that prevents cookies to be set for cross-domain requests in the browser. Additionally, it should be stated that this flag has no effect in non-browser environments.

nondeterministic committed:
First round of review clean-ups

nondeterministic committed:
Some more docu

nondeterministic committed:
Fix

hmil commented:
Almost there. This will help you fix the tests:
1. Add `preflightContinue: true` to the `cors` middleware configuration in the nodejs server. This is what causes the `Method` tests to fail.
2. There is a test that consists in making sure the examples given in the README compile. Update that test using the script `test/update-readme-sanity-check.sh`, or manually apply the diff [given here](https://travis-ci.org/hmil/RosHTTP/jobs/447471426#L1208) to `shared/src/test/scala/fr/hmil/roshttp/StreamingPressureTest.scala`.

nondeterministic committed:
Add preflight to cors-module and README tests to test spec

nondeterministic commented:
Thanks for all your patience and guidance - very much appreciated! The JVM/JS tests now all run fine, however, the Firefox tests still have some method/header issues it seems, which are not fixed by allowing preflight:

```
[info] fr.hmil.roshttp.HttpRequestSpec.Request headers.Overwrite previous value when set
[info]                                                  utest.AssertionError: assert(res.body.contains("\"cache-control\":\"max-age=128\""))
[info]                                                  res: fr.hmil.roshttp.response.SimpleHttpResponse = fr.hmil.roshttp.response.SimpleHttpResponse@13
[info] fr.hmil.roshttp.HttpRequestSpec.Request headers.Override body content-type               Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request headers          Success
[info] fr.hmil.roshttp.HttpRequestSpec.Response headers.can be read in the general case
[info]                                          java.util.NoSuchElementException: key not found: X-Powered-By
[info] fr.hmil.roshttp.HttpRequestSpec.Response headers.can be read in the error case
[info]                                          java.util.NoSuchElementException: key not found: X-Powered-By
[info] fr.hmil.roshttp.HttpRequestSpec.Response headers         Success
[info] fr.hmil.roshttp.HttpRequestSpec.Http method.can be set to any legal value
[info]                                          java.util.NoSuchElementException: key not found: X-Request-Method
[info] fr.hmil.roshttp.HttpRequestSpec.Http method.ignores case and capitalizes
[info]                                          java.util.NoSuchElementException: key not found: X-Request-Method
[info] fr.hmil.roshttp.HttpRequestSpec.Http method              Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request body.Plain text.works with ASCII strings         Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request body.Plain text.works with non-ASCII strings             Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request body.Plain text          Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request body.Multipart.works as intended         Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request body.Multipart           Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request body.URL encoded.works as intended               Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request body.URL encoded         Success
[info] fr.hmil.roshttp.HttpRequestSpec.Request body.JSON.works as intended
[info]                                  fr.hmil.roshttp.exceptions.RequestException: A network error occurred during HTTP request transmission.
```

Note that the specific CORS test passes without problems. So I am not quite sure how serious/unexpected the above observation is. Other than that, it seems that this last push to the PR satisfied the CI again.

Also I assumed in my push to the PR that instead of `StreamingPressureTest.scala` you meant `ReadmeSanityCheck.scala`. `StreamingPressureTest.scala` seems to contain nothing that would be affected by what I've changed... Do let me know if this was wrong, please.

P.S. I have not committed anything to increase your project's version number. I can do it, but I did not know what number increase you'd like to make.

hmil commented:
Sorry for the delay. I didn't realize I had disabled browser-based tests in CI.

The reason the browser tests are failing is because the cors middleware you've added to the test server sets a strict policy and the browser trips on security features.

You need to expose the custom headers we use for testing, with [Access-Control-Expose-Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers). Then you also need to allow all methods we test for with [Access-Control-Allow-Methods](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods).

I'll merge your work and fix this on master. Thank you for your contribution!

nondeterministic commented:
I have to thank you! I learned quite a bit in the process.

hmil commented:
@nondeterministic I can't publish the updated package. The setup I used to publish is broken and I couldn't make it work within the time I allowed myself to spend on this.

I'm not invested enough in Scala these days to spend any longer trying to fix this. As a matter of fact, the heaviness and over-engineering around build/packaging systems and tooling in the Scala ecosystem is the primary reason I'm moving away from it and towards TypeScript, even though Scala in itself is a superior programming language.

As stated many times, this library is looking for a maintainer. I'm sorry I don't have the energy to keep the boat afloat.

nondeterministic commented:
@hmil, is there a way that I could try and fix the release? Could we try it together? I'm sure you agree, that right now I'm not in a good position to take over maintenance of such a comprehensive library, but I'm willing to lend a hand.  Also, I'd find it mighty sad, if this useful feature didn't make it into a release. If you like, we can move this discussion to email - mine is visible in my profile. Cheers!

hmil commented:
Hold on, I'm trying another way.

hmil commented:
@nondeterministic Give that a shot:
Set the roshttp version to 2.2.3, and add this to your `build.sbt` file: `resolvers += Resolver.bintrayRepo("hmil", "roshttp")`

Let me know if that works for you.

nondeterministic commented:
I think, this should be `Resolver.bintrayRepo("hmil", "maven")` instead. Otherwise the package is not found. But with this modification, it seems to work just fine.

hmil commented:
Awesome! I've updated the instructions in the readme. I'll keep publishing this package to bintray in the future.
